File: typescript/packages/mechanisms/evm/src/erc4337/paymaster.ts
/**
 * ERC-4337 Paymaster Client
 *
 * Handles paymaster interactions for gas sponsorship.
 * Supports verifying paymasters (off-chain signature) and
 * sponsoring paymasters (third-party gas payment).
 */

import type { Address, Hex } from "viem";
import { concat, pad, toHex, keccak256, encodeAbiParameters } from "viem";
import type {
  UserOperation,
  PaymasterData,
  PaymasterConfig,
  GasEstimate,
} from "./types.js";
import { DEFAULT_GAS_LIMITS } from "./constants.js";

/**
 * Paymaster service response
 */
export interface PaymasterResponse {
  /** Paymaster address */
  paymaster: Address;
  /** Paymaster data to include in UserOp */
  paymasterData: Hex;
  /** Gas limits for paymaster operations */
  paymasterVerificationGasLimit: bigint;
  paymasterPostOpGasLimit: bigint;
}

/**
 * Paymaster sponsor request
 */
export interface SponsorRequest {
  /** UserOperation to sponsor (without paymaster data) */
  userOp: Partial<UserOperation>;
  /** Chain ID */
  chainId: number;
  /** EntryPoint address */
  entryPoint: Address;
  /** Optional context for the paymaster */
  context?: Record<string, unknown>;
}

/**
 * Paymaster client for gas sponsorship
 */
export class PaymasterClient {
  private readonly config: PaymasterConfig;

  constructor(config: PaymasterConfig) {
    this.config = config;
  }

  /**
   * Get paymaster data for a UserOperation
   */
  async getPaymasterData(
    userOp: Partial<UserOperation>,
    chainId: number,
    entryPoint: Address,
    context?: Record<string, unknown>,
  ): Promise<PaymasterData> {
    switch (this.config.type) {
      case "verifying":
        return this.getVerifyingPaymasterData(userOp, chainId, entryPoint);
      case "sponsoring":
        return this.getSponsoringPaymasterData(
          userOp,
          chainId,
          entryPoint,
          context,
        );
      case "token":
        return this.getTokenPaymasterData(userOp, chainId, entryPoint);
      default:
        throw new Error(`Unknown paymaster type: ${this.config.type}`);
    }
  }

  /**
   * Get gas estimates including paymaster gas
   */
  async estimatePaymasterGas(
    userOp: Partial<UserOperation>,
    _chainId: number,
  ): Promise<GasEstimate> {
    // For most paymasters, use default gas limits
    // This can be overridden by calling the paymaster service
    return {
      verificationGasLimit: DEFAULT_GAS_LIMITS.verificationGasLimit,
      callGasLimit: DEFAULT_GAS_LIMITS.callGasLimit,
      preVerificationGas: DEFAULT_GAS_LIMITS.preVerificationGas,
      paymasterVerificationGasLimit:
        DEFAULT_GAS_LIMITS.paymasterVerificationGasLimit,
      paymasterPostOpGasLimit: DEFAULT_GAS_LIMITS.paymasterPostOpGasLimit,
    };
  }

  /**
   * Check if the paymaster will sponsor this operation
   */
  async willSponsor(
    userOp: Partial<UserOperation>,
    chainId: number,
    entryPoint: Address,
    context?: Record<string, unknown>,
  ): Promise<boolean> {
    if (!this.config.url) {
      // Local paymaster - always sponsors
      return true;
    }

    try {
      const response = await fetch(`${this.config.url}/check`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          userOp: this.serializeUserOp(userOp),
          chainId,
          entryPoint,
          context,
        }),
      });

      if (!response.ok) return false;

      const result = (await response.json()) as { willSponsor: boolean };
      return result.willSponsor;
    } catch {
      return false;
    }
  }

  /**
   * Get verifying paymaster data (off-chain signature)
   */
  private async getVerifyingPaymasterData(
    userOp: Partial<UserOperation>,
    chainId: number,
    entryPoint: Address,
  ): Promise<PaymasterData> {
    if (this.config.url) {
      // Call paymaster service for signature
      return this.callPaymasterService(userOp, chainId, entryPoint);
    }

    // Local verifying paymaster - return basic data
    // The signature would need to be added by the paymaster owner
    return {
      paymaster: this.config.address,
      paymasterVerificationGasLimit:
        DEFAULT_GAS_LIMITS.paymasterVerificationGasLimit,
      paymasterPostOpGasLimit: DEFAULT_GAS_LIMITS.paymasterPostOpGasLimit,
      paymasterData: "0x" as Hex,
    };
  }

  /**
   * Get sponsoring paymaster data (third-party pays)
   */
  private async getSponsoringPaymasterData(
    userOp: Partial<UserOperation>,
    chainId: number,
    entryPoint: Address,
    context?: Record<string, unknown>,
  ): Promise<PaymasterData> {
    if (!this.config.url) {
      throw new Error("Sponsoring paymaster requires a service URL");
    }

    // Call the sponsor API
    const response = await fetch(`${this.config.url}/sponsor`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        userOp: this.serializeUserOp(userOp),
        chainId,
        entryPoint,
        context,
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Paymaster rejected sponsorship: ${error}`);
    }

    const result = (await response.json()) as PaymasterResponse;

    return {
      paymaster: result.paymaster,
      paymasterVerificationGasLimit: result.paymasterVerificationGasLimit,
      paymasterPostOpGasLimit: result.paymasterPostOpGasLimit,
      paymasterData: result.paymasterData,
    };
  }

  /**
   * Get token paymaster data (pay gas with ERC20)
   */
  private async getTokenPaymasterData(
    userOp: Partial<UserOperation>,
    chainId: number,
    entryPoint: Address,
  ): Promise<PaymasterData> {
    const tokenAddress = this.config.options?.tokenAddress as Address | undefined;
    if (!tokenAddress) {
      throw new Error("Token paymaster requires tokenAddress in options");
    }

    if (this.config.url) {
      // Call paymaster service for token rate and data
      return this.callPaymasterService(userOp, chainId, entryPoint, {
        tokenAddress,
      });
    }

    // Return basic token paymaster data
    // The actual rate and validation would be done on-chain
    return {
      paymaster: this.config.address,
      paymasterVerificationGasLimit:
        DEFAULT_GAS_LIMITS.paymasterVerificationGasLimit,
      paymasterPostOpGasLimit: DEFAULT_GAS_LIMITS.paymasterPostOpGasLimit,
      paymasterData: tokenAddress as Hex, // Token address as data
    };
  }

  /**
   * Call paymaster service API
   */
  private async callPaymasterService(
    userOp: Partial<UserOperation>,
    chainId: number,
    entryPoint: Address,
    context?: Record<string, unknown>,
  ): Promise<PaymasterData> {
    if (!this.config.url) {
      throw new Error("Paymaster service URL not configured");
    }

    const response = await fetch(`${this.config.url}/getPaymasterData`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        userOp: this.serializeUserOp(userOp),
        chainId,
        entryPoint,
        context,
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Paymaster service error: ${error}`);
    }

    const result = (await response.json()) as PaymasterResponse;

    return {
      paymaster: result.paymaster,
      paymasterVerificationGasLimit: BigInt(result.paymasterVerificationGasLimit),
      paymasterPostOpGasLimit: BigInt(result.paymasterPostOpGasLimit),
      paymasterData: result.paymasterData,
    };
  }

  /**
   * Serialize UserOperation for API calls
   */
  private serializeUserOp(
    userOp: Partial<UserOperation>,
  ): Record<string, string> {
    const result: Record<string, string> = {};

    if (userOp.sender) result.sender = userOp.sender;
    if (userOp.nonce !== undefined)
      result.nonce = `0x${userOp.nonce.toString(16)}`;
    if (userOp.initCode) result.initCode = userOp.initCode;
    if (userOp.callData) result.callData = userOp.callData;
    if (userOp.verificationGasLimit !== undefined)
      result.verificationGasLimit = `0x${userOp.verificationGasLimit.toString(16)}`;
    if (userOp.callGasLimit !== undefined)
      result.callGasLimit = `0x${userOp.callGasLimit.toString(16)}`;
    if (userOp.preVerificationGas !== undefined)
      result.preVerificationGas = `0x${userOp.preVerificationGas.toString(16)}`;
    if (userOp.maxPriorityFeePerGas !== undefined)
      result.maxPriorityFeePerGas = `0x${userOp.maxPriorityFeePerGas.toString(16)}`;
    if (userOp.maxFeePerGas !== undefined)
      result.maxFeePerGas = `0x${userOp.maxFeePerGas.toString(16)}`;
    if (userOp.paymasterAndData) result.paymasterAndData = userOp.paymasterAndData;
    if (userOp.signature) result.signature = userOp.signature;

    return result;
  }
}

/**
 * Create a PaymasterClient instance
 */
export function createPaymasterClient(config: PaymasterConfig): PaymasterClient {
  return new PaymasterClient(config);
}

/**
 * Encode paymaster data for inclusion in UserOperation
 */
export function encodePaymasterAndData(data: PaymasterData): Hex {
  return concat([
    data.paymaster,
    pad(toHex(data.paymasterVerificationGasLimit), { size: 16 }),
    pad(toHex(data.paymasterPostOpGasLimit), { size: 16 }),
    data.paymasterData,
  ]) as Hex;
}

/**
 * Decode paymaster and data from UserOperation
 */
export function decodePaymasterAndData(paymasterAndData: Hex): PaymasterData | null {
  if (paymasterAndData === "0x" || paymasterAndData.length < 86) {
    return null;
  }

  // 20 bytes address + 16 bytes verification gas + 16 bytes postOp gas = 52 bytes = 104 hex chars + 0x
  const paymaster = `0x${paymasterAndData.slice(2, 42)}` as Address;
  const paymasterVerificationGasLimit = BigInt(
    `0x${paymasterAndData.slice(42, 74)}`,
  );
  const paymasterPostOpGasLimit = BigInt(`0x${paymasterAndData.slice(74, 106)}`);
  const paymasterData = `0x${paymasterAndData.slice(106)}` as Hex;

  return {
    paymaster,
    paymasterVerificationGasLimit,
    paymasterPostOpGasLimit,
    paymasterData,
  };
}

File: go/mechanisms/evm/erc4337/paymaster.go
package erc4337

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"time"

	"github.com/ethereum/go-ethereum/common"
)

// PimlicoPaymasterConfig contains configuration for Pimlico paymaster.
type PimlicoPaymasterConfig struct {
	// APIKey is the Pimlico API key
	APIKey string `json:"apiKey"`
	// ChainID is the chain ID
	ChainID int64 `json:"chainId"`
	// PaymasterURL is an optional custom paymaster URL
	PaymasterURL string `json:"paymasterUrl,omitempty"`
	// EntryPoint is the EntryPoint contract address
	EntryPoint common.Address `json:"entryPoint,omitempty"`
	// SponsorshipPolicyID is the optional sponsorship policy ID
	SponsorshipPolicyID string `json:"sponsorshipPolicyId,omitempty"`
}

// TokenQuote contains quote information for token paymaster.
type TokenQuote struct {
	// Token address
	Token common.Address `json:"token"`
	// Symbol
	Symbol string `json:"symbol"`
	// Decimals
	Decimals int `json:"decimals"`
	// Fee in token
	Fee *big.Int `json:"fee"`
	// Exchange rate
	ExchangeRate *big.Int `json:"exchangeRate"`
}

// PimlicoPaymaster is a Pimlico paymaster client.
type PimlicoPaymaster struct {
	apiKey              string
	chainID             int64
	paymasterURL        string
	entryPoint          common.Address
	sponsorshipPolicyID string
	httpClient          *http.Client
	requestID           int
}

// NewPimlicoPaymaster creates a new Pimlico paymaster client.
func NewPimlicoPaymaster(config PimlicoPaymasterConfig) *PimlicoPaymaster {
	paymasterURL := config.PaymasterURL
	if paymasterURL == "" {
		network := getPimlicoNetwork(config.ChainID)
		paymasterURL = fmt.Sprintf("https://api.pimlico.io/v2/%s/rpc?apikey=%s", network, config.APIKey)
	}

	entryPoint := config.EntryPoint
	if entryPoint == (common.Address{}) {
		entryPoint = common.HexToAddress(EntryPointV07Address)
	}

	return &PimlicoPaymaster{
		apiKey:              config.APIKey,
		chainID:             config.ChainID,
		paymasterURL:        paymasterURL,
		entryPoint:          entryPoint,
		sponsorshipPolicyID: config.SponsorshipPolicyID,
		httpClient:          &http.Client{Timeout: 30 * time.Second},
	}
}

// SponsorUserOperation sponsors a UserOperation.
func (p *PimlicoPaymaster) SponsorUserOperation(userOp *UserOperation) (*PaymasterData, error) {
	packed := packUserOpForPaymaster(userOp)

	params := []interface{}{packed, p.entryPoint.Hex()}
	if p.sponsorshipPolicyID != "" {
		params = append(params, map[string]string{
			"sponsorshipPolicyId": p.sponsorshipPolicyID,
		})
	}

	var result struct {
		PaymasterAndData              string `json:"paymasterAndData"`
		Paymaster                     string `json:"paymaster,omitempty"`
		PaymasterVerificationGasLimit string `json:"paymasterVerificationGasLimit,omitempty"`
		PaymasterPostOpGasLimit       string `json:"paymasterPostOpGasLimit,omitempty"`
		PaymasterData                 string `json:"paymasterData,omitempty"`
	}

	err := p.rpcCall("pm_sponsorUserOperation", params, &result)
	if err != nil {
		return nil, err
	}

	paymasterData := &PaymasterData{}

	// Try v0.7 format first (separate fields)
	if result.Paymaster != "" {
		paymasterData.Paymaster = common.HexToAddress(result.Paymaster)
		if result.PaymasterVerificationGasLimit != "" {
			paymasterData.PaymasterVerificationGasLimit = hexToBigInt(result.PaymasterVerificationGasLimit)
		}
		if result.PaymasterPostOpGasLimit != "" {
			paymasterData.PaymasterPostOpGasLimit = hexToBigInt(result.PaymasterPostOpGasLimit)
		}
		if result.PaymasterData != "" {
			paymasterData.PaymasterData = hexToBytes(result.PaymasterData)
		}
	} else if result.PaymasterAndData != "" && result.PaymasterAndData != "0x" {
		// Fall back to v0.6 format (packed)
		data := hexToBytes(result.PaymasterAndData)
		if len(data) >= 20 {
			paymasterData.Paymaster = common.BytesToAddress(data[:20])
		}
		if len(data) > 20 {
			paymasterData.PaymasterData = data[20:]
		}
	}

	return paymasterData, nil
}

// GetTokenQuotes gets quotes for paying gas with tokens.
func (p *PimlicoPaymaster) GetTokenQuotes(userOp *UserOperation, tokens []common.Address) ([]TokenQuote, error) {
	packed := packUserOpForPaymaster(userOp)

	tokenStrings := make([]string, len(tokens))
	for i, token := range tokens {
		tokenStrings[i] = token.Hex()
	}

	var result []struct {
		Token        string `json:"token"`
		Symbol       string `json:"symbol"`
		Decimals     int    `json:"decimals"`
		Fee          string `json:"fee"`
		ExchangeRate string `json:"exchangeRate"`
	}

	err := p.rpcCall("pimlico_getTokenQuotes", []interface{}{
		packed,
		p.entryPoint.Hex(),
		tokenStrings,
	}, &result)
	if err != nil {
		return nil, err
	}

	quotes := make([]TokenQuote, len(result))
	for i, r := range result {
		quotes[i] = TokenQuote{
			Token:        common.HexToAddress(r.Token),
			Symbol:       r.Symbol,
			Decimals:     r.Decimals,
			Fee:          hexToBigInt(r.Fee),
			ExchangeRate: hexToBigInt(r.ExchangeRate),
		}
	}

	return quotes, nil
}

// GetPaymasterData implements PaymasterClient interface.
func (p *PimlicoPaymaster) GetPaymasterData(userOp *UserOperation, chainID int64, entryPoint common.Address) (*PaymasterData, error) {
	return p.SponsorUserOperation(userOp)
}

// WillSponsor checks if the paymaster will sponsor this operation.
func (p *PimlicoPaymaster) WillSponsor(userOp *UserOperation, chainID int64, entryPoint common.Address) (bool, error) {
	_, err := p.SponsorUserOperation(userOp)
	return err == nil, nil
}

func (p *PimlicoPaymaster) rpcCall(method string, params []interface{}, result interface{}) error {
	p.requestID++

	request := map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      p.requestID,
		"method":  method,
		"params":  params,
	}

	body, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	resp, err := p.httpClient.Post(p.paymasterURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP error %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Result json.RawMessage `json:"result"`
		Error  *struct {
			Code    int             `json:"code"`
			Message string          `json:"message"`
			Data    json.RawMessage `json:"data,omitempty"`
		} `json:"error"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return fmt.Errorf("failed to decode response: %w", err)
	}

	if response.Error != nil {
		return fmt.Errorf("RPC error %d: %s", response.Error.Code, response.Error.Message)
	}

	if result != nil && len(response.Result) > 0 {
		if err := json.Unmarshal(response.Result, result); err != nil {
			return fmt.Errorf("failed to unmarshal result: %w", err)
		}
	}

	return nil
}

// BiconomyPaymasterConfig contains configuration for Biconomy paymaster.
type BiconomyPaymasterConfig struct {
	// APIKey is the Biconomy API key
	APIKey string `json:"apiKey"`
	// ChainID is the chain ID
	ChainID int64 `json:"chainId"`
	// PaymasterURL is the paymaster URL
	PaymasterURL string `json:"paymasterUrl"`
	// Mode is the paymaster mode: "sponsored" or "erc20"
	Mode string `json:"mode"`
}

// BiconomyPaymaster is a Biconomy paymaster client.
type BiconomyPaymaster struct {
	apiKey       string
	chainID      int64
	paymasterURL string
	mode         string
	httpClient   *http.Client
	requestID    int
}

// NewBiconomyPaymaster creates a new Biconomy paymaster client.
func NewBiconomyPaymaster(config BiconomyPaymasterConfig) *BiconomyPaymaster {
	return &BiconomyPaymaster{
		apiKey:       config.APIKey,
		chainID:      config.ChainID,
		paymasterURL: config.PaymasterURL,
		mode:         config.Mode,
		httpClient:   &http.Client{Timeout: 30 * time.Second},
	}
}

// GetPaymasterData gets paymaster data for a UserOperation.
func (p *BiconomyPaymaster) GetPaymasterData(userOp *UserOperation, chainID int64, entryPoint common.Address) (*PaymasterData, error) {
	packed := packUserOpForPaymaster(userOp)

	request := map[string]interface{}{
		"method":        "pm_sponsorUserOperation",
		"userOperation": packed,
		"entryPoint":    entryPoint.Hex(),
		"chainId":       chainID,
		"mode":          p.mode,
	}

	var result struct {
		PaymasterAndData              string `json:"paymasterAndData"`
		Paymaster                     string `json:"paymaster,omitempty"`
		PaymasterVerificationGasLimit string `json:"paymasterVerificationGasLimit,omitempty"`
		PaymasterPostOpGasLimit       string `json:"paymasterPostOpGasLimit,omitempty"`
		PaymasterData                 string `json:"paymasterData,omitempty"`
	}

	err := p.rpcCall("pm_sponsorUserOperation", []interface{}{request}, &result)
	if err != nil {
		return nil, err
	}

	paymasterData := &PaymasterData{}
	if result.Paymaster != "" {
		paymasterData.Paymaster = common.HexToAddress(result.Paymaster)
		if result.PaymasterVerificationGasLimit != "" {
			paymasterData.PaymasterVerificationGasLimit = hexToBigInt(result.PaymasterVerificationGasLimit)
		}
		if result.PaymasterPostOpGasLimit != "" {
			paymasterData.PaymasterPostOpGasLimit = hexToBigInt(result.PaymasterPostOpGasLimit)
		}
		if result.PaymasterData != "" {
			paymasterData.PaymasterData = hexToBytes(result.PaymasterData)
		}
	} else if result.PaymasterAndData != "" && result.PaymasterAndData != "0x" {
		data := hexToBytes(result.PaymasterAndData)
		if len(data) >= 20 {
			paymasterData.Paymaster = common.BytesToAddress(data[:20])
		}
		if len(data) > 20 {
			paymasterData.PaymasterData = data[20:]
		}
	}

	return paymasterData, nil
}

// WillSponsor checks if the paymaster will sponsor this operation.
func (p *BiconomyPaymaster) WillSponsor(userOp *UserOperation, chainID int64, entryPoint common.Address) (bool, error) {
	_, err := p.GetPaymasterData(userOp, chainID, entryPoint)
	return err == nil, nil
}

// GetFeeQuotes gets fee quotes for ERC20 token payment.
func (p *BiconomyPaymaster) GetFeeQuotes(userOp *UserOperation, tokens []common.Address) ([]TokenQuote, error) {
	packed := packUserOpForPaymaster(userOp)

	tokenStrings := make([]string, len(tokens))
	for i, token := range tokens {
		tokenStrings[i] = token.Hex()
	}

	var result []struct {
		Token        string `json:"token"`
		Symbol       string `json:"symbol"`
		Decimals     int    `json:"decimals"`
		Fee          string `json:"fee"`
		ExchangeRate string `json:"exchangeRate"`
	}

	err := p.rpcCall("pm_getFeeQuotes", []interface{}{
		packed,
		tokenStrings,
	}, &result)
	if err != nil {
		return nil, err
	}

	quotes := make([]TokenQuote, len(result))
	for i, r := range result {
		quotes[i] = TokenQuote{
			Token:        common.HexToAddress(r.Token),
			Symbol:       r.Symbol,
			Decimals:     r.Decimals,
			Fee:          hexToBigInt(r.Fee),
			ExchangeRate: hexToBigInt(r.ExchangeRate),
		}
	}

	return quotes, nil
}

func (p *BiconomyPaymaster) rpcCall(method string, params []interface{}, result interface{}) error {
	p.requestID++

	request := map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      p.requestID,
		"method":  method,
		"params":  params,
	}

	body, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", p.paymasterURL, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", p.apiKey)

	resp, err := p.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP error %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Result json.RawMessage `json:"result"`
		Error  *struct {
			Code    int             `json:"code"`
			Message string          `json:"message"`
			Data    json.RawMessage `json:"data,omitempty"`
		} `json:"error"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return fmt.Errorf("failed to decode response: %w", err)
	}

	if response.Error != nil {
		return fmt.Errorf("RPC error %d: %s", response.Error.Code, response.Error.Message)
	}

	if result != nil && len(response.Result) > 0 {
		if err := json.Unmarshal(response.Result, result); err != nil {
			return fmt.Errorf("failed to unmarshal result: %w", err)
		}
	}

	return nil
}

// StackupPaymasterConfig contains configuration for Stackup paymaster.
type StackupPaymasterConfig struct {
	// APIKey is the Stackup API key
	APIKey string `json:"apiKey"`
	// ChainID is the chain ID
	ChainID int64 `json:"chainId"`
	// PaymasterURL is the paymaster URL
	PaymasterURL string `json:"paymasterUrl"`
	// Type is the paymaster type
	Type string `json:"type,omitempty"`
}

// StackupPaymaster is a Stackup paymaster client.
type StackupPaymaster struct {
	apiKey       string
	chainID      int64
	paymasterURL string
	pmType       string
	httpClient   *http.Client
	requestID    int
}

// NewStackupPaymaster creates a new Stackup paymaster client.
func NewStackupPaymaster(config StackupPaymasterConfig) *StackupPaymaster {
	return &StackupPaymaster{
		apiKey:       config.APIKey,
		chainID:      config.ChainID,
		paymasterURL: config.PaymasterURL,
		pmType:       config.Type,
		httpClient:   &http.Client{Timeout: 30 * time.Second},
	}
}

// GetPaymasterData gets paymaster data for a UserOperation.
func (p *StackupPaymaster) GetPaymasterData(userOp *UserOperation, chainID int64, entryPoint common.Address) (*PaymasterData, error) {
	packed := packUserOpForPaymaster(userOp)

	context := map[string]interface{}{}
	if p.pmType != "" {
		context["type"] = p.pmType
	}

	var result struct {
		PaymasterAndData              string `json:"paymasterAndData"`
		Paymaster                     string `json:"paymaster,omitempty"`
		PaymasterVerificationGasLimit string `json:"paymasterVerificationGasLimit,omitempty"`
		PaymasterPostOpGasLimit       string `json:"paymasterPostOpGasLimit,omitempty"`
		PaymasterData                 string `json:"paymasterData,omitempty"`
	}

	err := p.rpcCall("pm_getPaymasterStubData", []interface{}{
		packed,
		entryPoint.Hex(),
		fmt.Sprintf("0x%x", chainID),
		context,
	}, &result)
	if err != nil {
		return nil, err
	}

	paymasterData := &PaymasterData{}
	if result.Paymaster != "" {
		paymasterData.Paymaster = common.HexToAddress(result.Paymaster)
		if result.PaymasterVerificationGasLimit != "" {
			paymasterData.PaymasterVerificationGasLimit = hexToBigInt(result.PaymasterVerificationGasLimit)
		}
		if result.PaymasterPostOpGasLimit != "" {
			paymasterData.PaymasterPostOpGasLimit = hexToBigInt(result.PaymasterPostOpGasLimit)
		}
		if result.PaymasterData != "" {
			paymasterData.PaymasterData = hexToBytes(result.PaymasterData)
		}
	} else if result.PaymasterAndData != "" && result.PaymasterAndData != "0x" {
		data := hexToBytes(result.PaymasterAndData)
		if len(data) >= 20 {
			paymasterData.Paymaster = common.BytesToAddress(data[:20])
		}
		if len(data) > 20 {
			paymasterData.PaymasterData = data[20:]
		}
	}

	return paymasterData, nil
}

// WillSponsor checks if the paymaster will sponsor this operation.
func (p *StackupPaymaster) WillSponsor(userOp *UserOperation, chainID int64, entryPoint common.Address) (bool, error) {
	_, err := p.GetPaymasterData(userOp, chainID, entryPoint)
	return err == nil, nil
}

func (p *StackupPaymaster) rpcCall(method string, params []interface{}, result interface{}) error {
	p.requestID++

	request := map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      p.requestID,
		"method":  method,
		"params":  params,
	}

	body, err := json.Marshal(request)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", p.paymasterURL, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+p.apiKey)

	resp, err := p.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP error %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Result json.RawMessage `json:"result"`
		Error  *struct {
			Code    int             `json:"code"`
			Message string          `json:"message"`
			Data    json.RawMessage `json:"data,omitempty"`
		} `json:"error"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return fmt.Errorf("failed to decode response: %w", err)
	}

	if response.Error != nil {
		return fmt.Errorf("RPC error %d: %s", response.Error.Code, response.Error.Message)
	}

	if result != nil && len(response.Result) > 0 {
		if err := json.Unmarshal(response.Result, result); err != nil {
			return fmt.Errorf("failed to unmarshal result: %w", err)
		}
	}

	return nil
}

// Helper function to pack UserOperation for paymaster
func packUserOpForPaymaster(userOp *UserOperation) map[string]interface{} {
	result := map[string]interface{}{
		"sender":   userOp.Sender.Hex(),
		"callData": bytesToHex(userOp.CallData),
	}

	if userOp.Nonce != nil {
		result["nonce"] = bigIntToHex(userOp.Nonce)
	} else {
		result["nonce"] = "0x0"
	}

	if len(userOp.InitCode) > 0 {
		result["initCode"] = bytesToHex(userOp.InitCode)
	} else {
		result["initCode"] = "0x"
	}

	if userOp.VerificationGasLimit != nil {
		result["verificationGasLimit"] = bigIntToHex(userOp.VerificationGasLimit)
	} else {
		result["verificationGasLimit"] = bigIntToHex(DefaultGasLimits.VerificationGasLimit)
	}

	if userOp.CallGasLimit != nil {
		result["callGasLimit"] = bigIntToHex(userOp.CallGasLimit)
	} else {
		result["callGasLimit"] = bigIntToHex(DefaultGasLimits.CallGasLimit)
	}

	if userOp.PreVerificationGas != nil {
		result["preVerificationGas"] = bigIntToHex(userOp.PreVerificationGas)
	} else {
		result["preVerificationGas"] = bigIntToHex(DefaultGasLimits.PreVerificationGas)
	}

	if userOp.MaxFeePerGas != nil {
		result["maxFeePerGas"] = bigIntToHex(userOp.MaxFeePerGas)
	} else {
		result["maxFeePerGas"] = "0x0"
	}

	if userOp.MaxPriorityFeePerGas != nil {
		result["maxPriorityFeePerGas"] = bigIntToHex(userOp.MaxPriorityFeePerGas)
	} else {
		result["maxPriorityFeePerGas"] = "0x0"
	}

	if len(userOp.PaymasterAndData) > 0 {
		result["paymasterAndData"] = bytesToHex(userOp.PaymasterAndData)
	} else {
		result["paymasterAndData"] = "0x"
	}

	if len(userOp.Signature) > 0 {
		result["signature"] = bytesToHex(userOp.Signature)
	} else {
		result["signature"] = getDummySignature()
	}

	return result
}
